src/scheduler/pbs_sched.c

```diff
@@ -161,30 +164,37 @@ extern char *msg_startup1;
 void
 on_segv(int sig)
 {
-	int *thid;
+	int ret_lock = -1;
 
-	thid = (int *) pthread_getspecific(th_id_key);
-	if (thid != NULL && *thid != 0)
+	/* We want any other threads to block here, we want them alive until abort() is called
+	 * as it dumps core for all threads
+	 */
+	ret_lock = pthread_mutex_lock(&cleanup_lock);
+	if (ret_lock != 0)
 		pthread_exit(NULL);
 
-	if (num_threads > 1)
-		kill_threads();
-
 	/* we crashed less then 5 minutes ago, lets not restart ourself */
 	if ((segv_last_time - segv_start_time) < 300) {
-		log_record(PBSEVENT_SYSTEM, PBS_EVENTCLASS_SERVER, LOG_INFO, "on_segv", "received a sigsegv within 5 minutes of start: aborting.");
+		log_record(PBSEVENT_SYSTEM, PBS_EVENTCLASS_SERVER, LOG_INFO, __func__,
+				"received a sigsegv within 5 minutes of start: aborting.");
 		abort();
+
+		/* This will never really be executed, but it'll make sure that only 1 thread calls abort */
+		pthread_mutex_unlock(&cleanup_lock);
 	}
 
-	log_record(PBSEVENT_SYSTEM, PBS_EVENTCLASS_SERVER, LOG_INFO, "on_segv", "received segv and restarting");
+	log_record(PBSEVENT_SYSTEM, PBS_EVENTCLASS_SERVER, LOG_INFO, __func__,
+			"received segv and restarting");
 
 	if (fork() > 0) { /* the parent rexec's itself */
-		sleep(10);		/* allow the child to die */
+		sleep(10); /* allow the child to die */
+		pthread_mutex_unlock(&cleanup_lock);
```

Comment:
I don't think you need unlock in either case. This will ensure that no thread is context switched before execv is called.